query-planner-js/src/composedSchema/__tests__/removeInaccessibleElements.test.ts

```diff
@@ -0,0 +1,253 @@
+import { buildSchema, assertValidSchema } from 'graphql';
+import { removeInaccessibleElements } from '../removeInaccessibleElements';
+
+describe('removeInaccessibleElements', () => {
+  it(`removes @inaccessible fields`, () => {
+    let schema = buildSchema(`
+      directive @core(feature: String!) repeatable on SCHEMA
+
+      directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION
+
+      schema
+        @core(feature: "https://specs.apollo.dev/core/v0.1")
+        @core(feature: "https://specs.apollo.dev/inaccessible/v0.1")
+      {
+        query: Query
+      }
+
+      type Query {
+        someField: String
+        privateField: String @inaccessible
+      }
+    `);
+
+    schema = removeInaccessibleElements(schema);
+
+    const queryType = schema.getQueryType()!;
+
+    expect(queryType.getFields()['someField']).toBeDefined();
+    expect(queryType.getFields()['privateField']).toBeUndefined();
+  });
+
+  it(`removes @inaccessible object types`, () => {
+    let schema = buildSchema(`
+      directive @core(feature: String!) repeatable on SCHEMA
+
+      directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION
+
+      schema
+        @core(feature: "https://specs.apollo.dev/core/v0.1")
+        @core(feature: "https://specs.apollo.dev/inaccessible/v0.1")
+      {
+        query: Query
+      }
+
+      type Query {
+        fooField: Foo @inaccessible
+      }
+
+      type Foo @inaccessible {
+        someField: String
+      }
+
+      union Bar = Foo
+    `);
+
+    schema = removeInaccessibleElements(schema);
+
+    expect(schema.getType('Foo')).toBeUndefined();
+  });
+
+  it(`removes @inaccessible interface types`, () => {
+    let schema = buildSchema(`
+      directive @core(feature: String!) repeatable on SCHEMA
+
+      directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION
+
+      schema
+        @core(feature: "https://specs.apollo.dev/core/v0.1")
+        @core(feature: "https://specs.apollo.dev/inaccessible/v0.1")
+      {
+        query: Query
+      }
+
+      type Query {
+        fooField: Foo @inaccessible
+      }
+
+      interface Foo @inaccessible {
+        someField: String
+      }
+
+      type Bar implements Foo {
+        someField: String
+      }
+    `);
+
+    schema = removeInaccessibleElements(schema);
+
+    expect(schema.getType('Foo')).toBeUndefined();
```

Comment:
Maybe also add an expectation that `type Bar` exists and has `someField`, but doesn't implement any interfaces?

Reply:
Good suggestion, added in f950a4b